binary-option/program/src/processor.rs

```diff
@@ -0,0 +1,725 @@
+    {
+        binary_option.winning_side_pubkey = *winning_mint_account_info.key;
+    } else {
+        return Err(BinaryOptionError::InvalidWinner.into());
+    }
+    binary_option.settled = true;
+    binary_option.serialize(&mut *binary_option_account_info.data.borrow_mut())?;
+    Ok(())
+}
+
+pub fn process_collect(program_id: &Pubkey, accounts: &[AccountInfo]) -> ProgramResult {
+    msg!("Collect");
+    let account_info_iter = &mut accounts.iter();
+    let binary_option_account_info = next_account_info(account_info_iter)?;
+    let collector_info = next_account_info(account_info_iter)?;
+    let collector_long_token_account_info = next_account_info(account_info_iter)?;
+    let collector_short_token_account_info = next_account_info(account_info_iter)?;
+    let collector_account_info = next_account_info(account_info_iter)?;
+    let long_token_mint_info = next_account_info(account_info_iter)?;
+    let short_token_mint_info = next_account_info(account_info_iter)?;
+    let escrow_account_info = next_account_info(account_info_iter)?;
+    let escrow_authority_info = next_account_info(account_info_iter)?;
+    let fee_payer_info = next_account_info(account_info_iter)?;
+    let token_program_info = next_account_info(account_info_iter)?;
+    let system_account_info = next_account_info(account_info_iter)?;
+    let rent_info = next_account_info(account_info_iter)?;
+
+    let collector_long_token_account: Account =
+        assert_initialized(collector_long_token_account_info)?;
+    let collector_short_token_account: Account =
+        assert_initialized(collector_short_token_account_info)?;
+    let collector_account: Account = assert_initialized(collector_account_info)?;
+    let escrow_account: Account = assert_initialized(escrow_account_info)?;
+    let mut binary_option = BinaryOption::try_from_slice(&binary_option_account_info.data.borrow_mut())?;
+
+    // Get program derived address for escrow
+    let (escrow_owner_key, bump_seed) = Pubkey::find_program_address(
+        &[
+            long_token_mint_info.key.as_ref(),
+            short_token_mint_info.key.as_ref(),
+            token_program_info.key.as_ref(),
+            program_id.as_ref(),
+        ],
+        program_id,
+    );
+    let seeds = &[
+        long_token_mint_info.key.as_ref(),
+        short_token_mint_info.key.as_ref(),
+        token_program_info.key.as_ref(),
+        program_id.as_ref(),
+        &[bump_seed],
+    ];
+
+    if !binary_option.settled {
+        return Err(BinaryOptionError::BetNotSettled.into());
+    }
+    assert_owned_by(long_token_mint_info, &spl_token::id())?;
+    assert_owned_by(short_token_mint_info, &spl_token::id())?;
+    assert_keys_equal(collector_long_token_account.owner, *collector_info.key)?;
+    assert_keys_equal(collector_short_token_account.owner, *collector_info.key)?;
+    assert_keys_equal(collector_account.owner, *collector_info.key)?;
```

Comment:
these types of checks are pretty restrictive, what's the danger of not providing the collector info?

Reply:
You can pass in valid a `collector_long_token_account` and `collector_short_token_account` owned by me and a `collector_account` owner by yourself, and you've just rugged me.

Reply:
Ah ok, I see now, since you don't know beforehand how the funds are going to be moved between the 6 accounts.  I thought the token program would catch it since there's going to be a bad signature somewhere